### PR TITLE
Issue/1442 add bind options

### LINKED ISSUE #1442: Add config option for listen address of the server
- [x] Add option to specify the listen address of the server
- [x] Make sure in tests the server only binds to localhost

References:

https://github.com/inmanta/inmanta/blob/a10576814beef840050e2cf306ac9a71c4df7858/src/inmanta/protocol/rest/server.py#L290

The current server address option is something else than the listen socket:
https://github.com/inmanta/inmanta/blob/a10576814beef840050e2cf306ac9a71c4df7858/src/inmanta/server/config.py#L148

Currently the listen port is defined in a "transport" block, which is a relic from the past. The proposal to address this:

- Introduce bind-address and bind-port options in the server section. 
- If one of these is set, the server_rest_transport is ignored, but with a clear warning in the logs.
- Use 127.0.0.1 and 8888 as the default values
- Add a clear warning in the changelog that by default we now listen on 127.0.0.1:8888 instead of 0.0.0.0:8888
- bind-address should support both ipv4 and ipv6.

### PR DESCRIPTION
**Pull request by arnaudsjs**
Fixes #1442

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,16 @@ v 2019.4 (?) Changes in this release:
 - Use python3 from the core OS in Dockerfile
 - Introduce v2 protocol and implement project and environment api in v2 (#1412)
 - Improve agent documentation (#1389)
+- Introduce the bind-address and bind-port config option (#1442)
+
+BREAKING CHANGES:
```
Comment (bartv) [inline CHANGELOG.md]:
```suggestion
IMPORTANT CHANGES:
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,16 @@ v 2019.4 (?) Changes in this release:
...
+* The Inmanta server now listens on 127.0.0.1:8888 by default, while
```
Comment (bartv) [inline CHANGELOG.md]:
```suggestion
- The Inmanta server now listens on 127.0.0.1:8888 by default, while
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,16 @@ v 2019.4 (?) Changes in this release:
...
+  this was 0.0.0.0:8888 in previous versions. This behavior is
+  configurable with the `bind-address` config option.
+  
+DEPRECATIONS:
+* The `server_rest_transport.port` config option is deprecated in favour
```
Comment (bartv) [inline CHANGELOG.md]:
```suggestion
- The `server_rest_transport.port` config option is deprecated in favor
```

File: src/inmanta/server/config.py [hunk 4]
```diff
@@ -63,6 +78,27 @@
...
+    "server", "bind-address", "127.0.0.1", "A list of addresses on which the server will listen for connections", is_list
+)
+server_bind_port = Option("server", "bind-port", 8888, "The port on which the server will listen for connections", is_int)
+
+
+def get_bind_port() -> int:
+    if Config.is_set("server", "bind-port") or Config.is_set("server", "bind-address"):
+        # Use new bind-port option
+        if Config.is_set("server_rest_transport", "port"):
+            LOGGER.warning(
```
Comment (bartv) [inline src/inmanta/server/config.py]: Lets use warnings for this and push them also to the logs
https://docs.python.org/3.6/library/warnings.html
https://docs.python.org/3.6/library/logging.html#logging.captureWarnings

File: tests/conftest.py [hunk 5]
```diff
@@ -480,7 +481,7 @@ def map_data_line(line):
     try:
         await ibl.start()
     except SliceStartupException as e:
-        port = config.Config.get("server_rest_transport", "port")
+        port = config.Config.get("server", "bind-port")
```
Comment (bartv) [inline tests/conftest.py]: These print statements are not shown in the logs. Use logger for these

File: src/inmanta/server/config.py [hunk 6]
```diff
@@ -63,6 +79,29 @@
 #############################
 # server
 #############################
+server_bind_address = Option(
+    "server", "bind-address", "127.0.0.1", "A list of addresses on which the server will listen for connections", is_list
+)
+server_bind_port = Option("server", "bind-port", 8888, "The port on which the server will listen for connections", is_int)
+
+
+def get_bind_port() -> int:
```
Comment (wouterdb) [inline src/inmanta/server/config.py]: why not use Option.predecessor_option
or alternatively, use the option_as_default wrapper 
or extend the option_as_default wrapper  to do deprecation? 

This is an exception to the way options are fetched

File: tests/test_config.py [hunk 7]
```diff
@@ -169,3 +179,143 @@ def test_configfile_hierarchy(tmpdir):
     assert Config.get("influxdb", "tags")["tag2"] == "value2"
     assert Config.get("dashboard", "path") == "/directory"
     assert Config.get("dashboard", "client-id") == "test456"
+
+
+@pytest.mark.asyncio
+async def test_bind_address_ipv4(async_finalizer, client):
```
Comment (wouterdb) [inline tests/test_config.py]: can use some explanation

**Pull request merged**